File: flask/blog/blog/data/segment_data.py
from typing import Any
from data import database


def segments_of_post(post_id: int) -> list[dict[str, Any]]:
    segments = database.find_all(
        "SELECT idsegment,post_idpost,type,text,source FROM public.segment WHERE post_idpost=%s;",
        [str(post_id)],
    )

    return list(
        map(
            lambda segment: {
                "idsegment": segment[0],
                "postIdpost": segment[1],
                "type": segment[2],
                "text": segment[3],
                "source": segment[4],
            },
            segments,
        )
    )


def segment_by_id(segment_id: int) -> Any:
    return database.find_single(
        "SELECT idsegment,post_idpost,type,text,source FROM public.segment WHERE idsegment=%s;",
        [str(segment_id)],
    )


def new_segment(post_id: int, type: str, text: str, source: str) -> dict[str, Any]:
    segment_id: int = database.insert_data_returning_id(
        """INSERT INTO public.segment (post_idpost,type,text,source)
        VALUES (%s,%s,%s,%s) RETURNING idsegment;""",
        [str(post_id), type, text, source],
    )

    segment = segment_by_id(segment_id)

    return {
        "idsegment": segment[0],
        "postIdpost": segment[1],
        "type": segment[2],
        "text": segment[3],
        "source": segment[4],
    }


def delete_segment(segment_id: int):
    database.delete_single_by_id(
        "DELETE FROM public.segment WHERE idsegment=%s;", segment_id
    )
    # TODO

File: spring/blog/src/main/java/development/blog/data/SegmentData.java
package development.blog.data;

import development.blog.logic.User;
import development.blog.representations.request.NewSegment;
import development.blog.representations.response.PostResponse;
import development.blog.representations.response.SegmentResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.sql.*;
import java.util.ArrayList;

@Service
public class SegmentData {
    static final Logger log = LoggerFactory.getLogger(User.class);
    Connection connection;
    CommonData commonData;

    public SegmentData(Database database, CommonData commonData) {
        this.connection = database.getConnection();
        this.commonData = commonData;
    }

    public ArrayList<SegmentResponse> segmentsOfPost(Long id) throws SQLException {
        final String sql = "SELECT idsegment,post_idpost,type,text,source FROM public.segment WHERE post_idpost=?;";
        PreparedStatement statement =
                this.connection.prepareStatement(sql);
        statement.setLong(1,id);
        ResultSet rs = statement.executeQuery();
        ArrayList<SegmentResponse> segments = new ArrayList<SegmentResponse>();

        while (rs.next()) {
            segments.add(new SegmentResponse(
                    rs.getLong(1),
                    rs.getLong(2),
                    rs.getString(3),
                    rs.getString(4),
                    rs.getString(5)
            ));
        }
        return segments;
    }

    public SegmentResponse addSegment(NewSegment newSegment) throws SQLException {
        final String sql = "INSERT INTO public.segment (post_idpost,type,text,source) VALUES (?,?,?,?);";
        PreparedStatement statement =
                this.connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        statement.setLong(1, newSegment.getPostIdPost());
        statement.setString(2, newSegment.getType());
        statement.setString(3, newSegment.getText());
        statement.setString(4, newSegment.getSource());

        statement.executeUpdate();

        ResultSet rs = statement.getGeneratedKeys();
        rs.next();

        Long generatedId = rs.getLong(1);

        return getSegmentById(generatedId);
    }

    public SegmentResponse getSegmentById(Long id) throws SQLException {
        final String sqlSegment = "SELECT idsegment,post_idpost,type,text,source FROM public.segment WHERE idsegment=?;";
        PreparedStatement segmentStatement =
                this.connection.prepareStatement(sqlSegment);
        segmentStatement.setLong(1, id);
        ResultSet segmentRs = segmentStatement.executeQuery();
        segmentRs.next();

        return new SegmentResponse(
                segmentRs.getLong(1),
                segmentRs.getLong(2),
                segmentRs.getString(3),
                segmentRs.getString(4),
                segmentRs.getString(5)
        );
    }

    public void deleteSegment(Long id) throws SQLException {
        commonData.deleteById(id, "segment", "idsegment");
    }
}
